akka-remote/src/main/scala/akka/remote/artery/Handshake.scala

```diff
@@ -191,16 +192,29 @@ private[akka] class InboundHandshake(inboundContext: InboundContext, inControlSt
           override def onPush(): Unit = {
             val env = grab(in)
             env.message match {
-              case HandshakeReq(from) ⇒ onHandshakeReq(from)
+              case HandshakeReq(from, to) ⇒ onHandshakeReq(from, to)
               case _ ⇒
                 onMessage(env)
             }
           }
         })
 
-      private def onHandshakeReq(from: UniqueAddress): Unit = {
-        after(inboundContext.completeHandshake(from)) {
-          inboundContext.sendControl(from.address, HandshakeRsp(inboundContext.localAddress))
+      private def onHandshakeReq(from: UniqueAddress, to: Address): Unit = {
+        if (to == inboundContext.localAddress.address) {
+          after(inboundContext.completeHandshake(from)) {
+            inboundContext.sendControl(from.address, HandshakeRsp(inboundContext.localAddress))
+            pull(in)
+          }
+        } else {
+          if (log.isWarningEnabled)
```

Comment:
nitpick: isWarningEnabled not needed since this should not happen often and there is nothing costly with the log parameter (no allocations)